packages/filesystem/src/browser/location/location-renderer.tsx

```diff
@@ -18,32 +18,191 @@ import URI from '@theia/core/lib/common/uri';
 import { LocationService } from './location-service';
 import { ReactRenderer } from '@theia/core/lib/browser/widgets/react-renderer';
 import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import { FileService } from '../file-service';
+import { DisposableCollection, Emitter } from '@theia/core/lib/common';
 
+interface AutoSuggestDataEvent {
+    input: string;
+    parent: string;
+    children: string[];
+}
+
+class ResolvedDirectoryCache {
+    protected pendingResolvedDirectories = new Map<string, { input: string, pendingDirectories: Promise<void> }>();
+    protected cachedDirectories = new Map<string, string[]>();
+
+    protected directoryResolvedEmitter = new Emitter<AutoSuggestDataEvent>();
+    readonly onDirectoryDidResolve = this.directoryResolvedEmitter.event;
+
+    constructor(protected readonly fileService: FileService) { }
+
+    tryResolveChildDirectories(inputAsURI: URI): string[] | undefined {
+        const parentDirectory = inputAsURI.path.dir.toString();
+        const input = inputAsURI.path.toString();
+        if (this.cachedDirectories.has(parentDirectory)) {
```

Comment:
Minor: rather than 
```
if (...has) {
    ...get
}
```
I prefer
```
got = ...get
if (got) {
    use got
}
```
Because doing `has...get` means hashing the same input twice. But can be a little more muddled.

Reply:
Good catch, I've cleaned these lines up accordingly. I was also able to remove the last `else if` statement since I'm no longer updating the input value.